### PR TITLE
changing operations to compute function

### PR DESCRIPTION
**Pull request by ashmeigh**
### Summary of Changes
This pull request implements a significant shift in how operations are executed within our project, addressing issue. the code is moving from a partial function execution style to adopting a `compute_function` method across various operations. This transition facilitates a clearer separation between the setup and execution phases of operations.

### Key Changes:
- Introduced a `compute_function` static method in multiple operation classes, standardising the execution approach. This method encapsulates the core logic for operation execution on a per-slice basis.
- Adjusted the operation execution framework to support the new `compute_function` method, ensuring all operations are compatible with this standardised approach.
- Phased out the use of partial functions for operation execution in favour of the more explicit and structured `compute_function`

### CODE REVIEW

File: mantidimaging/core/operations/rebin/rebin.py [hunk 1]
```diff
@@ -31,8 +30,8 @@ class RebinFilter(BaseFilter):
     filter_name = "Rebin"
     link_histograms = True
 
-    @staticmethod
-    def filter_func(images: ImageStack, rebin_param=0.5, mode=None, progress=None) -> ImageStack:
+    @classmethod
```
Comment (samtygier-stfc) [inline mantidimaging/core/operations/rebin/rebin.py:33]: change back to staticmethod

File: mantidimaging/core/operations/rebin/rebin.py [hunk 2]
```diff
@@ -44,31 +43,26 @@ def filter_func(images: ImageStack, rebin_param=0.5, mode=None, progress=None) -
...
-                              mode=mode,
-                              output_shape=empty_resized_data.array.shape[1:])
-        ps.execute(partial_func=f,
-                   arrays=[images.shared_array, empty_resized_data],
-                   num_operations=images.data.shape[0],
-                   msg="Applying Rebin",
-                   progress=progress)
-        images.shared_array = empty_resized_data
+        params = {'new_shape': new_shape, 'mode': mode}
+        ps.run_compute_func(cls.compute_function, images.data.shape[0], images.shared_array, params, progress)
```
Comment (samtygier-stfc) [inline mantidimaging/core/operations/rebin/rebin.py:56]: use `RebinFilter.compute_function`

File: mantidimaging/core/operations/rebin/rebin.py [hunk 3]
```diff
@@ -44,31 +43,26 @@ def filter_func(images: ImageStack, rebin_param=0.5, mode=None, progress=None) -
...
+            current_shape = images.data.shape[1:]
+            new_shape = (int(current_shape[0] * rebin_param), int(current_shape[1] * rebin_param))
         else:
-            param_valid = rebin_param > 0
-
-        if not param_valid:
-            raise ValueError('Rebin parameter must be greater than 0')
+            raise ValueError("Invalid type for rebin_param")
 
-        empty_resized_data = _create_reshaped_array(images, rebin_param)
```
Comment (samtygier-stfc) [inline mantidimaging/core/operations/rebin/rebin.py:57]: need to keep this

File: mantidimaging/core/operations/rebin/rebin.py [hunk 4]
```diff
@@ -44,31 +43,26 @@ def filter_func(images: ImageStack, rebin_param=0.5, mode=None, progress=None) -
...
+        params = {'new_shape': new_shape, 'mode': mode}
+        ps.run_compute_func(cls.compute_function, images.data.shape[0], images.shared_array, params, progress)
 
         return images
 
+    @staticmethod
+    def compute_function(i: int, array: np.ndarray, params: dict):
+        new_shape = params['new_shape']
+        mode = params['mode']
+        array[i] = resize(array[i], output_shape=new_shape, mode=mode, preserve_range=True)
```
Comment (samtygier-stfc) [inline mantidimaging/core/operations/rebin/rebin.py:64]:
```
    def compute_function(i: int, arrays: list[np.ndarray], params: dict):
        new_shape = params['new_shape']
        mode = params['mode']
        input_array, output_array = arrays
        output_array[i] = resize(input_array[i], output_shape=new_shape, mode=mode, preserve_range=True)
```

Comment (ashmeigh) [reply mantidimaging/core/operations/rebin/rebin.py:64]: AttributeError: 'numpy.ndarray' object has no attribute 'has_shared_memory'
i have changed the code to fix this

**Comment by samtygier-stfc** [comment]
Split into other PRs

**Pull request closed without merging**